src/ApplicationInsights.Kubernetes/TelemetryInitializers/KubernetesModule.cs

```diff
@@ -47,12 +50,15 @@ public static void Initialize(TelemetryConfiguration configuration, TimeSpan? ti
         /// </summary>
         /// <param name="loggerFactory"></param>
         /// <param name="timeout"></param>
-        public static void EnableKubernetes(IServiceCollection serviceCollection, TelemetryConfiguration configuration, TimeSpan? timeout = null)
+        public static void EnableKubernetes(IServiceCollection serviceCollection,
+            TelemetryConfiguration configuration,
+            TimeSpan? timeout = null,
+            bool useStub = false)
```

Comment:
In general, it's always better to minimize of test specific code in product code.
The most common practice is abstracting with interfaces. Instead of bool useStub, it's better to just accept a IK8sEnvironmentFactory as a parameter defaulting to null. Then create a production default one if it's null. Then the rest of your class can be freed from test specific code like "if (useStub) { "

Reply:
@brahmnes , thanks for the review. Great thinking.

Here are some of my perspectives:
Minimize test specific code in production code is very reasonable because we don't want any test code interference the production code. At the best, we don't have test code in production.

However, given the advantage to debug the code without setting up a real K8s environment, that is a compromise I am trying to make here.

Now that we have stub class v.s. the real classes, the risk of misuse shows up. 
And the way to ensure correct service is injected is: to introduce a switch explicitly. That switch could be:
1. IK8sEnvironmentFactory or
2. a boolean
I intentionally had chosen 2 here because: we have 2 stubs needed instead of 1:
* IKubeHttpClientSettingsProvider
* IK8sEnvironmentFactory
And when stubs are used, they are always applied coupled, that is why a single boolean is picked to control the use of the stubs or not. Here we also are not considering allowing 3rd party implementations being injected.

Now that we have picked the boolean as the way, to limit the number of possible point of failures become important. And the place recognized is the injector, where all services got registered. Once the correct service is registered, it will not become wrong. That is where one and only one **if-else** happens.